### PR TITLE
Add increment code needed for 3DVar

### PR DESCRIPTION
**Pull request by frld**
## Description

This adds some of the increments class methods needed for data assimilation (DA) to orca-jedi and updates the increments unit test to test this. **NB this is continuation of PR #85

This PR doesn't including all the increments methods needed for DA in particular I left out the (netcdf) writing code, I'll put this in a later PR. I have also attempted to minimise the changes to other parts of the code in this PR. I'm hoping this means it is not too big to review.

FYI my draft list for PRs below
* Changes to increment (this PR)
* Changes to state, nemo field writer/reader
* Changes to interpolator / add dirac test (delta function test)
* Implement Saber / error covariance training  including changes to geometry needed for this.
* 3dvar executable/test
* +++ Multiprocessing / Nemovar / Multiple variables / Linear balance / orca025ext / mpi  ....

Currently this code only works for single processor runs. I was hoping to leave implementing multiprocessing for a future PR.

Also I am aware that both the increments methods and state methods operate on atlas fieldsets so there is the possibility of calling generic methods for both increment and state using a generic fields class. I note this approach is used by lfric and also oops/l95 and probably other code in JEDI. Let me know if you have any thoughts on this. It might be nice to do this in orcajedi too, but ideally not this PR. I think it would be helpful to get the increment test (along with my initial code) in place first.

## Issue(s) addressed

N/A

## Dependencies

N/A

## Impact

There is no impact on other code as far I am aware.

## Checklist

- [x] I have updated the unit tests to cover the change
- [x] New functions are documented briefly via Doxygen comments in the code
- [x] I have linted my code using cpplint (I've run the coding_norms test and it passes)
- [x] I have run the unit tests (all the orca-jedi ctests have run and passed)
- [x] I have run mo-bundle to check integration with the rest of JEDI and run the unit tests under all environments

### CODE REVIEW

**Comment by frld** [comment]
A new mo-bundle test is now running see http://fcm1/cylc-review/cycles/frld/?suite=mo-bundle-orca-jedi-increment3

File: src/orca-jedi/increment/Increment.cc [hunk 1]
```diff
@@ -25,11 +26,14 @@
 
 #include "ufo/GeoVaLs.h"
 
-#include "orca-jedi/errorcovariance/ErrorCovariance.h"
```
Comment (twsearle) [inline src/orca-jedi/increment/Increment.cc:28]: Do we still need this error covariance class at all? I am not up with the state of the art on JEDI covariances. Are you using one from saber? I am guessing if the class was removed entirely I would need to make changes to the traits file.

File: src/orca-jedi/increment/Increment.h [hunk 2]
```diff
@@ -8,6 +8,7 @@
 #include <string>
 #include <memory>
 #include <vector>
+#include <tuple>
```
Comment (twsearle) [inline src/orca-jedi/increment/Increment.h:11]: I think you aren't using tuple any more, but I might have missed it.

File: src/orca-jedi/increment/Increment.cc [hunk 3]
```diff
@@ -38,14 +42,19 @@ namespace orcamodel {
 Increment::Increment(const Geometry & geom,
                      const oops::Variables & vars,
                      const util::DateTime & time)
-  : geom_(new Geometry(geom)), vars_(), time_(time),
+  : geom_(new Geometry(geom)), vars_(vars), time_(time),
     incrementFields_()
 {
```
Comment (twsearle) [inline src/orca-jedi/increment/Increment.cc:47]: I think this class wouldn't work with the fields distributed across more than one processor (the norm calculation for example would be unique to each processor). I think it is good to throw an exception in the increment constructor if there is more than 1 mpi process. Something like the following (untested):
```suggestion
{
  if (geom_->.getComm().size() != 1) {
    throw eckit::NotImplemented("orcamodel::Increment::Increment: Cannot construct an Increment with more than one MPI process.");
  }
```

**Review by twsearle — CHANGES_REQUESTED** [review]
Its all looking good, just a couple of discussion points, and a question.

File: src/orca-jedi/increment/Increment.cc (hunk 1 above)
Comment (frld) [reply src/orca-jedi/increment/Increment.cc:28]: Yes this code is probably not needed as I've been encouraged to do the error covariance work with SABER. I was planning to delete it in number 4 of my pull request plan where I implement the SABER covariance. I can make sure the traits are still working then also.

File: src/orca-jedi/increment/Increment.h (hunk 2 above)
Comment (frld) [reply src/orca-jedi/increment/Increment.h:11]: You're right - removed.

File: src/orca-jedi/state/State.h [hunk 4]
```diff
@@ -105,6 +105,8 @@ class State : public util::Printable,
   const oops::Variables & variables() const {return vars_;}
   oops::Variables & variables() {return vars_;}
 
+  atlas::Field getField(int) const;
```
Comment (twsearle) [inline src/orca-jedi/state/State.h:108]: Sorry, I just realised one more thing - this one won't be tested. Could you please add a test to `src/tests/orca-jedi/test_state.cc` just to retrieve a field and check its name matches what is expected?  I can help with this if needed.

**Review by twsearle — CHANGES_REQUESTED** [review]
sorry spotted one more thing

File: src/orca-jedi/state/State.h (hunk 4 above)
Comment (frld) [reply src/orca-jedi/state/State.h:108]: I'll give it a go and let you know if I'm stumped.

**Comment by frld** [comment]
> A new mo-bundle test is now running see http://fcm1/cylc-review/cycles/frld/?suite=mo-bundle-orca-jedi-increment3

These all passed except for a few lfric tests which run out of wallclock. This is a known issue unrelated to my developments.

**Review by twsearle — APPROVED** [review]
Looks good and passes the tests, thanks for the contribution!

**Pull request merged**